Docker daemon crashes after a sequence of API calls

After a sequence of API calls related to execs Docker daemon crashes with a panic.

`docker version`:
```
% docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   a34a1d5
 Built:        Sat Nov 21 00:48:57 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      linux/amd64
```

`docker info`:
```
% docker info
Containers: 2
Images: 72
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 76
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.13-boot2docker
Operating System: Boot2Docker 1.9.1 (TCL 6.4.1); master : cef800b - Fri Nov 20 19:33:59 UTC 2015
CPUs: 1
Total Memory: 996.2 MiB
Name: vdocker
ID: AX4W:Z2PE:OHDD:SMJK:PH5C:U4GR:WVNK:5YFF:N476:W6NW:NAKK:2KSM
Debug mode (server): true
 File Descriptors: 13
 Goroutines: 19
 System Time: 2016-01-15T14:08:27.509022642Z
 EventsListeners: 0
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

However, I've also reproduced this problem on an Amazon Linux VM (can't access it now), also with Docker 1.9.1.

`uname -a`:

```
% uname -a
Darwin qplex.local 14.5.0 Darwin Kernel Version 14.5.0: Tue Sep  1 21:23:09 PDT 2015; root:xnu-2782.50.1~1/RELEASE_X86_64 x86_64
% docker-machine ssh vdocker uname -a
Linux vdocker 4.1.13-boot2docker #1 SMP Fri Nov 20 19:05:50 UTC 2015 x86_64 GNU/Linux
```

Steps to reproduce:

1\. Create a simple container which does nothing:
```
% docker run -d ubuntu:latest bash -c "while true; do sleep 10s; done"
d243ec78bdc3b61655dbc8da993c662b462ac9b8fb478655023130f8d62c113d
```

2\. Invoke an API call to create an exec with a simple command:
```
% http --verify=no --cert $HOME/.docker/machine/machines/vdocker/cert.pem --cert-key $HOME/.docker/machine/machines/vdocker/key.pem POST https://192.168.99.100:2376/containers/d243ec78bdc3b61655dbc8da993c662b462ac9b8fb478655023130f8d62c113d/exec AttachStdin:=false AttachStdout:=true AttachStderr:=true Tty:=false Cmd:='["test", "-e", "/createme"]'
HTTP/1.1 201 Created
Content-Length: 74
Content-Type: application/json
Date: Fri, 15 Jan 2016 14:12:40 GMT
Server: Docker/1.9.1 (linux)

{
    "Id": "2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e"
}
```

3\. Start the newly created exec through the API:
```
% http --verify=no --cert $HOME/.docker/machine/machines/vdocker/cert.pem --cert-key $HOME/.docker/machine/machines/vdocker/key.pem POST https://192.168.99.100:2376/exec/2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e/start Detach:=false Tty:=false
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.raw-stream
```
Docker appears intact:
```
% docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
d243ec78bdc3        ubuntu:latest       "bash -c 'while true;"   2 minutes ago       Up 2 minutes                            hopeful_visvesvaraya
```

4\. Start the same command again:
```
% http --verify=no --cert $HOME/.docker/machine/machines/vdocker/cert.pem --cert-key $HOME/.docker/machine/machines/vdocker/key.pem POST https://192.168.99.100:2376/exec/2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e/start Detach:=false Tty:=false
HTTP/1.1 200 OK
Content-Type: application/vnd.docker.raw-stream
```
Docker is now down:
```
% docker ps
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```

Log file contains the following messages:
```
time="2016-01-15T14:14:05.080977771Z" level=debug msg="Calling POST /exec/2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e/start"
time="2016-01-15T14:14:05.081603096Z" level=info msg="POST /exec/2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e/start"
time="2016-01-15T14:14:05.081852263Z" level=debug msg="starting exec command 2c2c3dcc47bdd91f038b054a86faf3671fb2e08ce82ce74f4848b57068d3b77e in container d243ec78bdc3b61655dbc8da993c662b462ac9b8fb4786550
time="2016-01-15T14:14:05.081982354Z" level=debug msg="attach: stdout: begin"
time="2016-01-15T14:14:05.082009897Z" level=debug msg="attach: stderr: begin"
panic: close of closed channel
```

Full log is available [here](https://gist.github.com/netvl/d4f9d11a1ccc81b5bc19).

Naturally, I expected Docker not to crash.

This scenario actually worked some time ago, but I can only say for certain that it stopped working around a month ago.

--- a/daemon/exec.go
+++ b/daemon/exec.go
@@ -135,6 +135,11 @@ func (d *Daemon) ContainerExecStart(name string, stdin io.ReadCloser, stdout io.
 	}
 
 	ec.Lock()
+	if ec.ExitCode != nil {
+		ec.Unlock()
+		return derr.ErrorCodeExecExited.WithArgs(ec.ID)
+	}
+
 	if ec.Running {
 		ec.Unlock()
 		return derr.ErrorCodeExecRunning.WithArgs(ec.ID)
@@ -214,7 +219,7 @@ func (d *Daemon) Exec(c *container.Container, execConfig *exec.Config, pipes *ex
 		exitStatus = 128
 	}
 
-	execConfig.ExitCode = exitStatus
+	execConfig.ExitCode = &exitStatus
 	execConfig.Running = false
 
 	return exitStatus, err

--- a/daemon/exec/exec.go
+++ b/daemon/exec/exec.go
@@ -18,7 +18,7 @@ type Config struct {
 	*runconfig.StreamConfig
 	ID            string
 	Running       bool
-	ExitCode      int
+	ExitCode      *int
 	ProcessConfig *execdriver.ProcessConfig
 	OpenStdin     bool
 	OpenStderr    bool

--- a/errors/daemon.go
+++ b/errors/daemon.go
@@ -742,6 +742,15 @@ var (
 		HTTPStatusCode: http.StatusInternalServerError,
 	})
 
+	// ErrorCodeExecExited is generated when we try to start an exec
+	// but its already running.
+	ErrorCodeExecExited = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "EXECEXITED",
+		Message:        "Error: Exec command %s has already run",
+		Description:    "An attempt to start an 'exec' was made, but 'exec' was already run",
+		HTTPStatusCode: http.StatusConflict,
+	})
+
 	// ErrorCodeExecCantRun is generated when we try to start an exec
 	// but it failed for some reason.
 	ErrorCodeExecCantRun = errcode.Register(errGroup, errcode.ErrorDescriptor{

--- a/integration-cli/docker_api_exec_test.go
+++ b/integration-cli/docker_api_exec_test.go
@@ -8,6 +8,7 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"time"
 
 	"github.com/docker/docker/pkg/integration/checker"
 	"github.com/go-check/check"
@@ -66,33 +67,23 @@ func (s *DockerSuite) TestExecAPIStart(c *check.C) {
 	testRequires(c, DaemonIsLinux) // Uses pause/unpause but bits may be salvagable to Windows to Windows CI
 	dockerCmd(c, "run", "-d", "--name", "test", "busybox", "top")
 
-	startExec := func(id string, code int) {
-		resp, body, err := sockRequestRaw("POST", fmt.Sprintf("/exec/%s/start", id), strings.NewReader(`{"Detach": true}`), "application/json")
-		c.Assert(err, checker.IsNil)
-
-		b, err := readBody(body)
-		comment := check.Commentf("response body: %s", b)
-		c.Assert(err, checker.IsNil, comment)
-		c.Assert(resp.StatusCode, checker.Equals, code, comment)
-	}
-
 	id := createExec(c, "test")
-	startExec(id, http.StatusOK)
+	startExec(c, id, http.StatusOK)
 
 	id = createExec(c, "test")
 	dockerCmd(c, "stop", "test")
 
-	startExec(id, http.StatusNotFound)
+	startExec(c, id, http.StatusNotFound)
 
 	dockerCmd(c, "start", "test")
-	startExec(id, http.StatusNotFound)
+	startExec(c, id, http.StatusNotFound)
 
 	// make sure exec is created before pausing
 	id = createExec(c, "test")
 	dockerCmd(c, "pause", "test")
-	startExec(id, http.StatusConflict)
+	startExec(c, id, http.StatusConflict)
 	dockerCmd(c, "unpause", "test")
-	startExec(id, http.StatusOK)
+	startExec(c, id, http.StatusOK)
 }
 
 func (s *DockerSuite) TestExecAPIStartBackwardsCompatible(c *check.C) {
@@ -108,6 +99,30 @@ func (s *DockerSuite) TestExecAPIStartBackwardsCompatible(c *check.C) {
 	c.Assert(resp.StatusCode, checker.Equals, http.StatusOK, comment)
 }
 
+// #19362
+func (s *DockerSuite) TestExecAPIStartMultipleTimesError(c *check.C) {
+	dockerCmd(c, "run", "-d", "--name", "test", "busybox", "top")
+	execID := createExec(c, "test")
+	startExec(c, execID, http.StatusOK)
+
+	timeout := time.After(10 * time.Second)
+	var execJSON struct{ Running bool }
+	for {
+		select {
+		case <-timeout:
+			c.Fatal("timeout waiting for exec to start")
+		default:
+		}
+
+		inspectExec(c, execID, &execJSON)
+		if !execJSON.Running {
+			break
+		}
+	}
+
+	startExec(c, execID, http.StatusConflict)
+}
+
 func createExec(c *check.C, name string) string {
 	_, b, err := sockRequest("POST", fmt.Sprintf("/containers/%s/exec", name), map[string]interface{}{"Cmd": []string{"true"}})
 	c.Assert(err, checker.IsNil, check.Commentf(string(b)))
@@ -117,4 +132,23 @@ func createExec(c *check.C, name string) string {
 	}{}
 	c.Assert(json.Unmarshal(b, &createResp), checker.IsNil, check.Commentf(string(b)))
 	return createResp.ID
+}
+
+func startExec(c *check.C, id string, code int) {
+	resp, body, err := sockRequestRaw("POST", fmt.Sprintf("/exec/%s/start", id), strings.NewReader(`{"Detach": true}`), "application/json")
+	c.Assert(err, checker.IsNil)
+
+	b, err := readBody(body)
+	comment := check.Commentf("response body: %s", b)
+	c.Assert(err, checker.IsNil, comment)
+	c.Assert(resp.StatusCode, checker.Equals, code, comment)
+}
+
+func inspectExec(c *check.C, id string, out interface{}) {
+	resp, body, err := sockRequestRaw("GET", fmt.Sprintf("/exec/%s/json", id), nil, "")
+	c.Assert(err, checker.IsNil)
+	defer body.Close()
+	c.Assert(resp.StatusCode, checker.Equals, http.StatusOK)
+	err = json.NewDecoder(body).Decode(out)
+	c.Assert(err, checker.IsNil)
 }

